File: src-tauri/tests/domain/app_config.rs
use gui_lib::models::app_config::AppConfig;
use gui_lib::services::app_config::normalize_menu_keys;

#[test]
fn default_app_config_is_gui_schema() {
    let config = AppConfig::default();

    assert_eq!(config.schema_version, "gui.app.v1");
    assert_eq!(config.core.kernel, "zero");
    assert!(config.core.auto_connect);
    assert!(config.core.auto_start);
    assert_eq!(config.logs.level, "info");
    assert!(!config.ui.sidebar_collapsed);
    assert_eq!(config.ui.hidden_menu_keys, vec!["debug".to_string()]);
    assert!(config.ui.default_route.is_none());
    assert_eq!(config.local_proxy.host, "127.0.0.1");
    assert_eq!(config.local_proxy.port, 7890);
    assert!(config.local_proxy.source_proxy_config_id.is_none());
    assert!(config.routing.inject_common_rules);
}

#[test]
fn hidden_menu_keys_are_normalized() {
    let keys = normalize_menu_keys(vec![
        " core ".to_string(),
        "".to_string(),
        "logs".to_string(),
        "core".to_string(),
        "settings".to_string(),
    ]);

    assert_eq!(keys, vec!["core".to_string(), "logs".to_string()]);
}

#[test]
fn legacy_download_proxy_setting_is_ignored_and_not_persisted() {
    let mut value = serde_json::to_value(AppConfig::default()).unwrap();
    value["core"]["downloadProxyAuto"] = serde_json::Value::Bool(false);

    let config: AppConfig = serde_json::from_value(value).unwrap();
    let saved = serde_json::to_value(config).unwrap();

    assert!(saved["core"].get("downloadProxyAuto").is_none());
}

File: src/lib/types/app-config.ts
// Mirror of Rust models::app_config

import type { DnsConfig } from './dns';

export interface AppConfig {
  schemaVersion: string;
  core: AppCoreConfig;
  logs: AppLogConfig;
  ui: AppUiConfig;
  localProxy: AppLocalProxyConfig;
  tun: AppTunConfig;
  dns: AppDnsConfig;
  routing: AppRoutingConfig;
  urlTest: AppUrlTestConfig;
}

export interface AppCoreConfig {
  kernel: string;
  autoConnect: boolean;
  autoStart: boolean;
  cleanupProxyOnExit: boolean;
  executablePath?: string;
  downloadUrl?: string;
  configPath?: string;
  workingDir?: string;
  socket?: string;
  networkProbeUrls: string[];
}

export interface AppLogConfig {
  level: string;
  maxEntries: number;
}

export interface AppUiConfig {
  theme: string;        // "light" | "dark" | "system"
  uiMode: string;       // "lite" | "pro"
  sidebarCollapsed: boolean;
  hiddenMenuKeys: string[];
  trafficBallEnabled: boolean;
  defaultRoute?: string;
}

export interface AppLocalProxyConfig {
  host: string;
  port: number;
  sourceProxyConfigId?: string;
  bypass: string[];
}

export interface AppTunConfig {
  /** Explicit local desired state. Undefined preserves legacy Lite auto-connect until first toggle. */
  enabled?: boolean;
  name?: string;
  addr: string;
  mask: string;
  secondaryAddr?: string;
  tag: string;
  mtu: number;
  dualStack: boolean;
  dnsHijack: boolean;
}

export interface AppDnsConfig {
  enabled: boolean;
  config?: DnsConfig;
  dnsHijack: boolean;
}

export interface AppRoutingConfig {
  injectCommonRules: boolean;
}

export interface AppUrlTestConfig {
  toleranceMs: number;
}

export interface KernelSettingsExportResult {
  path: string;
  schemaVersion: string;
}

// Patch types for partial updates

export interface AppConfigPatch {
  core?: AppCoreConfigPatch;
  logs?: AppLogConfigPatch;
  ui?: AppUiConfigPatch;
  localProxy?: AppLocalProxyConfigPatch;
  tun?: AppTunConfigPatch;
  dns?: AppDnsConfigPatch;
  routing?: AppRoutingConfigPatch;
  urlTest?: AppUrlTestConfigPatch;
}

export interface AppCoreConfigPatch {
  kernel?: string;
  autoConnect?: boolean;
  autoStart?: boolean;
  cleanupProxyOnExit?: boolean;
  executablePath?: string | null;
  downloadUrl?: string | null;
  configPath?: string | null;
  workingDir?: string | null;
  socket?: string | null;
  networkProbeUrls?: string[];
}

export interface AppLogConfigPatch {
  level?: string;
  maxEntries?: number;
}

export interface AppUiConfigPatch {
  theme?: string;
  uiMode?: string;
  sidebarCollapsed?: boolean;
  hiddenMenuKeys?: string[];
  trafficBallEnabled?: boolean;
  defaultRoute?: string | null;
}

export interface AppLocalProxyConfigPatch {
  host?: string;
  port?: number;
  sourceProxyConfigId?: string | null;
  bypass?: string[];
}

export interface AppTunConfigPatch {
  enabled?: boolean;
  name?: string | null;
  addr?: string;
  mask?: string;
  secondaryAddr?: string | null;
  tag?: string;
  mtu?: number;
  dualStack?: boolean;
  dnsHijack?: boolean;
}

export interface AppDnsConfigPatch {
  enabled?: boolean;
  config?: DnsConfig | null;
  dnsHijack?: boolean;
}

export interface AppRoutingConfigPatch {
  injectCommonRules?: boolean;
}

export interface AppUrlTestConfigPatch {
  toleranceMs?: number;
}
